core/src/main/scala/org/apache/spark/executor/TaskMetrics.scala

```diff
@@ -287,11 +258,10 @@ private[spark] object TaskMetrics extends Logging {
   def fromAccumulatorUpdates(accumUpdates: Seq[AccumulableInfo]): TaskMetrics = {
     val definedAccumUpdates = accumUpdates.filter(_.update.isDefined)
     val metrics = new ListenerTaskMetrics(definedAccumUpdates)
-    // We don't register this [[ListenerTaskMetrics]] for cleanup, and this is only used to post
-    // event, we should un-register all accumulators immediately.
-    metrics.internalAccums.foreach(acc => Accumulators.remove(acc.id))
-    definedAccumUpdates.filter(_.internal).foreach { accum =>
-      metrics.internalAccums.find(_.name == accum.name).foreach(_.setValueAny(accum.update.get))
+    definedAccumUpdates.filter(_.internal).foreach { accInfo =>
+      metrics.internalAccums.find(_.name == accInfo.name).foreach { acc =>
+        acc.asInstanceOf[Accumulator[Any, Any]].add(accInfo.update.get)
```

Comment:
This is an example shows a weakness of the new API: we can't `setValue`.  For this example, we have the final output and we wanna set the value of accumulator so that it can produce the same output.  With the new API, we can't guarantee that all accumulators can implement `setValue`, e.g. the average accumulator.  I'm still thinking about how to fix it or work around it,  @rxin any ideas?